### ISSUE TITLE
Mi Kettle - We maybe also can find a way to get data from the MiKettle

### ISSUE THREAD

**Issue by TrevisGordan** [ThomDietrich/miflora-mqtt-daemon#22]
#Now MiKettle gets detected! (Xiomi Waterkettle)

Hey to Day I had to Fix something on the mqtt Deamon and suddenly My MiKettle was detected as a Bluetooth Device while running "lescan". 

I tried to conected but the MAC is restricted via:
`    if not re.match("C4:7C:8D:[0-9A-F]{2}:[0-9A-F]{2}:[0-9A-F]{2}", mac):`

so that did not work.
The Kettle uses Bluetooth as well and probably works similar to the flora sensors so at least the Water temp should be possible to fetch right?
Any Idea how to Check if this could also work?  

<img width="197" alt="bildschirmfoto 2018-02-09 um 22 29 09" src="https://user-images.githubusercontent.com/29541742/36051503-a48d26d4-0dea-11e8-8880-5bc840d5e0b1.png">

![1200x800](https://user-images.githubusercontent.com/29541742/36051573-d8b27540-0dea-11e8-82b0-65093ad20d47.jpg)

**Comment by tremez** [comment]
I have debugged it a bit, and it is possible to get data from Kettle. Unfortunately it requres some kind of authentication, and does not provide any data without authentication.

Anyone have experience with MiHome devices authentication ?

**Comment by ThomDietrich** [comment]
I've not used it so far but there is an openHAB binding for MiHome specifically. Check it out!

**Comment by aprosvetova** [comment]
Ta-da
https://github.com/aprosvetova/xiaomi-kettle

**Comment by tremez** [comment]
Great ! Thank you !
молодчинка

**Comment by drndos** [comment]
> Ta-da
> https://github.com/aprosvetova/xiaomi-kettle

Hello, we don't need perfect code 😄 , just working code, we can perfect it later. Could you please release the code? Thank you.

**Comment by aprosvetova** [comment]
@drndos, maybe later. Still need to refactor some things.

**Comment by beikeland** [comment]
> Ta-da

This is awesome, sad to see the having to physically press a button limitation is still a thing. Is there any work out there for a custom firmware, my searches turn up nothing?

**Comment by beikeland** [comment]
Thanks for the insight! Disassembly done; base contains power supply, relay and buzzer, temperature sensor.

Inside handle (under touch buttons) there is a small board with 2x 3pin connections to the base. 
(presumably power and relay + temperature and busser). Everything is encapsulated in clear silicone, some test points visible under the silicone. Reverse side is all plastic casing filled with silicone.

Will follow up if I get to removing the silicone. As the DT-06(?) module in the picture below indicates, a wifi version may be easier if the supply in the base can handle the load.

![image](https://user-images.githubusercontent.com/2701678/49442456-dba04100-f7c9-11e8-8482-882f78adf6a7.png)

**Comment by aprosvetova** [comment]
@beikeland of couse it’d be easier to replace the MCU with ESP32 or smth like that because you’ll have full access to firmware, but my goal was to hack the original kettle :D

**Comment by drndos** [comment]
> Thanks for the insight! Disassembly done; base contains power supply, relay and buzzer, temperature sensor.
> 
> Inside handle (under touch buttons) there is a small board with 2x 3pin connections to the base.
> (presumably power and relay + temperature and busser). Everything is encapsulated in clear silicone, some test points visible under the silicone. Reverse side is all plastic casing filled with silicone.
> 
> Will follow up if I get to removing the silicone. As the DT-06(?) module in the picture below indicates, a wifi version may be easier if the supply in the base can handle the load.
> 
> ![image](https://user-images.githubusercontent.com/2701678/49442456-dba04100-f7c9-11e8-8482-882f78adf6a7.png)

Replacing the bluetooth module with wifi one would quite possibly due to the size difference destroy the clean esthetics of the device. It seems like the PCB of the Mi Kettle is fully custom.

**Comment by drndos** [comment]
@aprosvetova Guys in dustcloud https://github.com/dgiese/dustcloud managed to hack firmware for xiaomi vacuum, maybe worth the research whether there is anything in common. Also some guy hacked miband https://medium.com/machine-learning-world/how-i-hacked-xiaomi-miband-2-to-control-it-from-linux-a5bd2f36d3ad  His tutorial seems very similar to yours. Maybe someone hacked firmware for miband and the procedure would be similar. I have managed to read stuff (not write) with his instructions and some BLE sniffing.

**Comment by beikeland** [comment]
Completely get that @aprosvetova , and there is likely a lot of people who would be very grateful. I figured I'd look into physical access in the off chance the chip was not locked for reading.

Getting to the board is simple. Three screws on the base (one under the barcode). To release the bezel on the handle, I depressed the locking tab inside, and gently pryed the gap that formed on the outside, the rest is snap fit, and one screw securing the board assembly.

@drndos There is ample room for at least a ESP8285 solution, quite possibly a ESP32 module as well. It would likely require a custom PCB to deal with the touch interface; but if there is no progress on the firmware I'll pursue that in the future. For now its just additional information.

**Comment by aprosvetova** [comment]
@drndos yea thanks, but they are completely different. Different chips, different communication ways, etc

**Comment by drndos** [comment]
@aprosvetova If we manage to extract the MCU and connect to its debug ports. We have two options either it is unprotected and ezpz but most probably locked. Do you have any experience in dumping protected program from MCU ? I tried some cold boot stepping with STM32, I am not sure if it would apply here. What would you suggest ?

**Comment by aprosvetova** [comment]
@drndos oh I’m more softwareish, I just have a little experience in microcontrollers. So if you dump the fw, I’ll be able to patch it, encrypt and pack as an Quintic OTA update and write some software to let us upload this update on the kettle.

**Comment by drndos** [comment]
I am reading through documentation on the MCU and it seems that flash is encrypted. But I see that the program is being loaded to sram on the start and I expect it to be unencrypted there. I know it is slightly different, but the logic might be viable: https://www.aisec.fraunhofer.de/en/FirmwareProtection.html
It seems like a long shot though, I don't have pirateswd on me nor do I have the means to do the precise power switching with my arduino. Do any of you have pirateswd ?

**Comment by drndos** [comment]
Also I have extracted "fc502ce05f9bebf032e9f32884390b23_upd_yunmi.kettle.v2.bin" is that the encrypted firmware ?

**Comment by aprosvetova** [comment]
Yes, it is, @drndos. I had different hash, maybe because of different version. Here's my bin
[fw.bin.zip](https://github.com/ThomDietrich/miflora-mqtt-daemon/files/2650547/fw.bin.zip)

**Comment by drndos** [comment]
@aprosvetova Thanks I found these parts are exactly same:
0000-0001
0022-0052
0082-0091
00B2-00C1
00F2-0341
0352-03C1
Nothing more so far :/

**Comment by aprosvetova** [comment]
We'd better go that MCU extraction way. I'm sure there is nothing interesting in these encrypted firmwares :(

**Comment by gsprod** [comment]
@drndos, @aprosvetova, any luck with your research ? any progress ? could you share what has been done so fare ? 
thanks, GS.

**Comment by Mattrees007** [comment]
@theanoni just wondering if you were able to convert your European kettle to Chinese?

**Comment by the4anoni** [comment]
> 
> 
> @theanoni just wondering if you were able to convert your European kettle to Chinese?

Not yet, we need to wait for devs :(

**Comment by cRaZy-bisCuiT** [comment]
Any news about the topic? :)

**Comment by drndos** [comment]
Hey guys, I made HA integration here: https://github.com/drndos/mikettle-ha
No integration for controlling the kettle for now though.

**Comment by cRaZy-bisCuiT** [comment]
> Hey guys, I made HA integration here: https://github.com/drndos/mikettle-ha
> No integration for controlling the kettle for now though.
But when I follow your links and and up at BLE protocol it seems like this in general is possible? Setting at least the heat up temperature would be very nice and all I need. :)

**Comment by johnpwhite** [comment]
> Anybody know how can i convert european version of kette to chinese so i can pair it with China Mainland server ?

I'm keen to switch from China to Australia, did you find a solution?

**Comment by ramikg** [comment]
>     2. decrypt firmware (how? Quintic provides us with encrypting software only)

Decryption is possible using [_qotaunpack_](https://github.com/ramikg/qotaunpack/) (given the key, of course).

**Comment by zackoch** [comment]
Yahtzee hotsauce
https://github.com/aleaksah/qn902x-dump
